File: wasm/bmxt-core/src/features/tabs_picker/new_window.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NewWindowTabMeta {
    pub id: i32,
    pub window_id: i32,
    pub index: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResolveNewWindowOrderContext {
    pub tabs: Vec<NewWindowTabMeta>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResolvedNewWindowOrder {
    pub ordered_ids: Vec<i32>,
}

pub fn resolve_new_window_order(ctx: ResolveNewWindowOrderContext) -> ResolvedNewWindowOrder {
    let mut tabs = ctx.tabs;
    tabs.sort_by(|a, b| {
        a.window_id
            .cmp(&b.window_id)
            .then_with(|| a.index.cmp(&b.index))
            .then_with(|| a.id.cmp(&b.id))
    });
    ResolvedNewWindowOrder {
        ordered_ids: tabs.into_iter().map(|t| t.id).collect(),
    }
}

File: lib/features/bmxt-core/tabs-picker/new-window.ts
export type NewWindowTabMeta = {
  id: number
  windowId: number
  index: number
}

export type ResolveNewWindowOrderContext = {
  tabs: NewWindowTabMeta[]
}

export type ResolvedNewWindowOrder = {
  orderedIds: number[]
}

export function resolveNewWindowOrder(ctx: ResolveNewWindowOrderContext): ResolvedNewWindowOrder {
  const tabs = [...ctx.tabs].sort((a, b) => {
    if (a.windowId !== b.windowId) return a.windowId - b.windowId
    if (a.index !== b.index) return a.index - b.index
    return a.id - b.id
  })
  return { orderedIds: tabs.map((t) => t.id) }
}

export function resolveTabsPickerNewWindowOrder<TContext, TResult>(context: TContext): TResult {
  return resolveNewWindowOrder(context as ResolveNewWindowOrderContext) as TResult
}
